internal/consensus/reactor_test.go

```diff
@@ -190,8 +190,14 @@ func waitForAndValidateBlock(
 	defer cancel()
 	fn := func(j int) {
 		msg, err := blocksSubs[j].Next(ctx)
-		if !assert.NoError(t, err) {
-			cancel()
+		switch {
+		case errors.Is(err, context.DeadlineExceeded):
+			return
+		case errors.Is(err, context.Canceled):
+			return
+		case err != nil:
+			t.Fatalf("problem waiting for %d subscription: %v", j, err)
+			cancel() // terminate other workers
```

Comment:
This is not reachable, t.Fatal does not return in the scope of the call. I think you can just swap them.

I believe deferring it would work, I believe it calls `runtime.Goexit()` so defers should happen.

Reply:
I think I had to fix this another way in the next patch because calling fatal in a goroutine is also verbotten.

Reply:
For the same reason, but from the other side. 🙂

Reply:
oops that was the other case, I think it's easier to just use testify in these cases,

Reply:
require also uses FailNow under the covers I think. Not that you shouldn't use it, but that alone wouldn't fix it.

Or if the idea is to just log a test failure this could be `t.Errorf`

Reply:
yeah, I think the nest of this is all in the "there are too many threads here," and the whole prospect of making an assertion inside of a thread is a big fraught.

Reply:
`assert` or `t.Error` should be safe anywhere—the restraints on `t.Fatal`/`require` are because of the goroutine exit handling.